A0-3083: Refactor ticker for chain extensions

timorl opened the pull request:
# Description

These are changes that are needed to use ticker with the periodic requests for chain extensions which we will be implementing soon. In particular this _should not_ change any current behaviour of the node, thus it's just a refactor.

Comments on whether this interface is clear or not are very welcome.

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have added tests
- I have made corresponding changes to the existing documentation
- I have created new documentation

kostekIV commented on finality-aleph/src/sync/ticker.rs:
```diff
@@ -1,39 +1,37 @@
-/// Example usage would be to use `wait` method in main select loop and `try_tick` whenever
-/// you would like to tick sooner in another branch of select.
+/// at most after `max_timeout` but no sooner than after `min_timeout`.
+/// It will always tick after `max_timeout` and never after that unless reset earlier.
```
```suggestion
/// It will always tick before `max_timeout` and never after that unless reset earlier.
```
?

timorl replied:
Ugh, that's also not what I meant, but trying to write what I meant turned out to bee too hard, so I just deleted this line.

The documentation next to the methods should be good enough.

kostekIV replied:
Yep, it was confusing as hell :P

kostekIV reviewed (commented):
Could you restore `wait_and_reset` fn for usecases where the user doesn't want to remember about resetting the timeout? This is a much less error prone approach. Why is the extraction of the reset is needed?

kostekIV commented on finality-aleph/src/sync/ticker.rs:
```diff
@@ -1,39 +1,36 @@
+/// resetting whenever the rate limited action actually occurs.
 pub struct Ticker {
-    last_tick: Instant,
+    last_reset: Instant,
```
while we are changing this could we also rename current_timeout to some better name :<

timorl commented:
The extraction of reset is needed because we will now have "failed ticks", i.e. the ticker ticks, but it turns out we don't actually want to send any message. We don't want to reset the ticker in this situation, since if anything suggests that we now want to send a message, we want to tick again immediately.

Making `wait` into `wait_and_reset` (or just keeping the name `wait_and_tick`) was something I even considered, the only issue I have with it now is that we either add weird boilerplate or always reset twice when we reach `max_timeout`. If you think the latter is acceptable then I can definitely change it, it only seemed somewhat annoying to me. To be clear, then we wouldn't even need a separate `wait` method – all the uses we have for this have a "forced" action when `wait` passes, so we would reset always anyway.

timorl commented:
Hm, changed significantly, especially since while addressing comments I noticed this wouldn't've been enough for the application it is supposed for. <_<"

Should be clearer what is happening now, although we have a double reset if `max_timeout` actually occurs.

kostekIV commented on finality-aleph/src/sync/ticker.rs:
```diff
@@ -1,13 +1,18 @@
 use tokio::time::{sleep, Duration, Instant};
 
+enum State {
```
`Mode` seems better here

timorl replied:
Done.

woocash2 reviewed (approved):
lgtm
